gapir/client/client.go

```diff
@@ -102,33 +140,152 @@ func (client *Client) Connect(ctx context.Context, device bind.Device, abi *devi
+			return
+		case <-time.After(heartbeatInterval):
+			err := replayer.ping(ctx)
+			if err != nil {
+				log.E(ctx, "Error sending keep-alive ping. Error: %v", err)
+				client.reconnect(ctx, replayer)
+				return
+			}
+		}
+	}
+}
+
+// getActiveReplayer returns the replayer identified by key only if this
+// replayer has an active connection to a GAPIR instance.
+func (client *Client) getActiveReplayer(ctx context.Context, key *ReplayerKey) (*replayerInfo, error) {
+	replayer, found := client.replayers[*key]
+	if !found {
+		return nil, log.Errf(ctx, nil, "Cannot find replayer for this key: %v", key)
+	}
+
+	if replayer.rpcClient == nil || replayer.conn == nil || replayer.rpcStream == nil {
+		return nil, log.Err(ctx, nil, "Replayer has no active connection")
+	}
+
+	return replayer, nil
+}
+
+// BeginReplay sends a replay request to the replayer identified by key.
+func (client *Client) BeginReplay(ctx context.Context, key *ReplayerKey, payload string, dependent string) error {
+	ctx = log.Enter(ctx, "Starting replay on gapir device")
+	replayerInfo, err := client.getActiveReplayer(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	idReq := replaysrv.ReplayRequest{
+		Req: &replaysrv.ReplayRequest_Replay{
+			Replay: &replaysrv.Replay{
+				ReplayId:    payload,
+				DependentId: dependent,
+			},
+		},
+	}
+	err = replayerInfo.rpcStream.Send(&idReq)
+	if err != nil {
+		return log.Err(ctx, err, "Sending replay id")
+	}
+
+	return nil
+}
+
+// SetReplayExecutor assigns a replay executor to the replayer identified by
+// key. It returns a cleanup function to remove the executor once the replay
+// is finished.
+func (client *Client) SetReplayExecutor(ctx context.Context, key *ReplayerKey, executor ReplayExecutor) (func(), error) {
+	replayerInfo, err := client.getActiveReplayer(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	if replayerInfo.executor != nil {
+		return nil, log.Err(ctx, nil, "Cannot set an executor while one is already present")
+	}
+	replayerInfo.executor = executor
+	return func() { replayerInfo.executor = nil }, nil
+}
 
+// PrewarmReplay requests the GAPIR device to get itself into the given state
+func (client *Client) PrewarmReplay(ctx context.Context, key *ReplayerKey, payload string, cleanup string) error {
+	replayerInfo, err := client.getActiveReplayer(ctx, key)
+	if err != nil {
+		return log.Err(ctx, err, "Getting replayer replayerInfo")
+	}
+
+	PrerunReq := replaysrv.ReplayRequest{
+		Req: &replaysrv.ReplayRequest_Prewarm{
+			Prewarm: &replaysrv.PrewarmRequest{
+				PrerunId:  payload,
+				CleanupId: cleanup,
+			},
+		},
+	}
+	err = replayerInfo.rpcStream.Send(&PrerunReq)
+	if err != nil {
+		return log.Err(ctx, err, "Sending replay payload")
+	}
+	return nil
+}
+
+// startReplayCommunicationHandler launches a background task which creates
+// the Replay RPC stream and starts to listen to it.
+func (replayer *replayerInfo) startReplayCommunicationHandler(ctx context.Context) error {
```

Comment:
I would prefer the definition of replayerInfo and the functions related with it in a different file then here both for readability and to reflect the actual structure.

Reply:
Agreed, moved to `replayer.go`